hie-plugin-api/Haskell/Ide/Engine/PluginApi.hs

```diff
@@ -0,0 +1,71 @@
+-- | This module provides an API that software intented to be
+-- integrated into HIE can use, so that they can make use of the
+-- shared BIOS features.
+
+{-
+-- Stuff used in HaRe currently
+Options(..)
+defaultOptions
+GmModuleGraph(..)
+ModulePath(..)
+GmComponent(..)
+GmComponentType(..)
+
+CachedInfo(..)
+HasGhcModuleCache(..)
+IdeGhcM
+
+cabalModuleGraphs
+filePathToUri
+makeRevRedirMapFunc
+
+MonadIO(..)
+ifCachedModule
+runIdeGhcMBare
+setTypecheckedModule
+-}
+
+
+module Haskell.Ide.Engine.PluginApi
+  (
+  -- ** Re-exported from ghc-mod via ghc-project-types
+    GP.GmModuleGraph(..)
+  , GP.ModulePath(..)
+  , GP.GmComponent(..)
+  , GP.GmComponentType(..)
+
+  -- * IDE monads
+  , HIE.IdeState(..)
+  , HIE.IdeGhcM
+  , HIE.runIdeGhcM
+  , HIE.runIdeGhcMBare
+  , HIE.IdeM
+  , HIE.runIdeM
+  , HIE.IdeDeferM
+  , HIE.MonadIde(..)
+  , HIE.iterT
+  , HIE.LiftsToGhc(..)
+  , HIE.HasGhcModuleCache(..)
+  , HIE.cabalModuleGraphs
+  , HIE.makeRevRedirMapFunc
```

Comment:
Is this meant to be a one-stop import for all plugins? If so should this also export the `withCachedModule` etc. functions and other helpers?

Reply:
@bubba the initial intention is to create as narrow as possible an API to support what HaRe needs, as a way to end up with a real world requirement on hie-bios.

And I think there possibly still needs to be a discussion about what the layers are for a plugin.

My mental model for an external tool XXX currently has it something like

- hie
- plugin within hie to intergrate XXX to hie/LSP
- XXX as a standalone tool
- part of hie exposed to interact with GHC etc, and allow the tool to be a good citizen in hie, but also to be able to run standalone.

So this file is currently aiming at the last of those, which is effectively the hie-bios. I think.

But this whole layering is something that needs discussion, I think.

Reply:
Perhaps we could view HIE as having two parts:
- A library for running long-lived sessions that provide access to processed information from GHC, in a relatively bulletproof fashion (i.e. using the hie-bios to work out how to talk to GHC)
    - Plugins could code against this library, and it would then also provide a way to run those tools in a relatively robust way. But they wouldn't have to have a dependency on a LSP client.
- A Haskell LSP client which uses the session library and manages plugins.

Reply:
@michaelpj Do you mean LSP server in your comment above? The client is the IDE that talks to hie.

Reply:
Er yes, sorry :man_facepalming:

Reply:
And the intended breakdown you describe is the driver behind having `hie-plugin-api` which should play the role described in your first bullet point, and does so for HaRe using this branch and https://github.com/alanz/HaRe/tree/hie-plugin-api.

But I can also envisage another layer, where a third party could package something like HaRe for hie, which can then be simply included in the plugin list in `app/MainHie.hs` and be usable.

So my original discussion point is whether we need to distinguish between the 'upper' and 'lower' APIs. Especially as I think there is likely to be some level of crossover in it.